### PR TITLE
feat(policy): support ListenerSet as targetRef for ClientTrafficPolicy

### LINKED ISSUE #8092: Support XListenerSet as TargetRefs for xPolicies
Related to https://github.com/envoyproxy/gateway/issues/5800

### PR DESCRIPTION
**Pull request by apkatsikas**
## What type of PR is this?

feature

## What this PR does / why we need it

Adds `ListenerSet` as a valid `targetRef` kind for `ClientTrafficPolicy`, alongside the existing `Gateway`.

Priority follows the existing more-specific-wins model: sectionName policies > ListenerSet-wide > Gateway-wide.

Changes:

- CEL validation: `kind == 'Gateway'` → `kind in ['Gateway', 'ListenerSet']`
- Translation: ListenerSet and Gateway targets share loop bodies; ListenerSet-wide is processed
  before Gateway-wide to enforce priority regardless of input order
- `claimedScopes` tracks whole-scope claims (Gateway-wide, ListenerSet-wide); `claimedSections` tracks individual listeners claimed by sectionName or ListenerSet-wide policies, keyed by a scope-qualified key (`lsPrefix + listenerName` for ListenerSet listeners, bare name for Gateway listeners) so that same-named listeners across different ListenerSets — or a Gateway listener sharing a name with a ListenerSet listener — do not produce false Conflicted/Overridden status
- Three new test fixtures cover the duplicate-name collision scenarios (wide and sectionName variants)
- CRDs regenerated (`make manifests`, `make helm-template.gateway-crds-helm`)

**Known gap / follow-up**: `ListenerSet` is missing from the upstream `sectionName` field description in `gwapiv1.LocalPolicyTargetReferenceWithSectionName`, [PR opened here](https://github.com/kubernetes-sigs/gateway-api/pull/4880). CEL rules in this PR enforce valid kinds at admission time regardless.

**Local verification**: KIND cluster with dev EG image, CTP with `tls.minVersion: "1.3"` targeting a ListenerSet HTTPS listener:

```powershell
# TLS 1.3 — succeeds
curl.exe -k --tlsv1.3 https://api.localhost
# Hello from Kubernetes

# TLS 1.2 — rejected by Envoy, proving the policy is enforced
curl.exe -k --tls-max 1.2 https://api.localhost
# curl: (35) schannel: ... SEC_E_UNSUPPORTED_FUNCTION
```

Gateway's own HTTP listener was unaffected — only the ListenerSet's HTTPS listener had the TLS restriction applied.

## Which issue(s) this PR fixes

#8092

## Release Notes

Yes

## Reviewing the diff

`clienttrafficpolicy.go` - to review only meaningful changes, append `?w=1` to the GitHub PR URL.

### CODE REVIEW

**Comment by jukie** [comment]
@codex review

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 1]
```diff
@@ -192,144 +200,171 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
+						errs                   error
+						http3DisabledListeners []string
+					)
+					for _, l := range gateway.listeners {
+						if policyTargetsLS && listenersDontBelongToLS(l, ls) {
+							continue
+						}
+						// Skip if section has already been targeted
+						if claimedSections[key].Has(string(l.Name)) {
+							continue
```
[the remark this answers is not in the document]
Comment (jukie) [reply internal/gatewayapi/clienttrafficpolicy.go]: Would be useful to include something that triggers this in the testdata example so that future changes to behavior are easier to spot.

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 2]
```diff
@@ -367,11 +451,15 @@ func resolveClientTrafficPolicyTargetRef(
...
-			return gateway.GatewayContext, err
+			return &ctpResolvedTarget{gateway: gateway.GatewayContext}, err
 		}
 	}
 
-	return gateway.GatewayContext, nil
+	return &ctpResolvedTarget{gateway: gateway.GatewayContext}, nil
+}
+
+func listenersDontBelongToLS(l *ListenerContext, ls *gwapiv1.ListenerSet) bool {
```
Comment (jukie) [inline internal/gatewayapi/clienttrafficpolicy.go]: Minor nit: can we flip this logic to `listenerBelongToLS`

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 3]
```diff
@@ -192,144 +200,171 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
+				gateways,
+				resources.ReferenceGrants,
+				egv1a1.GroupName,
+				egv1a1.KindClientTrafficPolicy,
+				currPolicy.Namespace,
+				t.GetNamespace,
+			)
+			for _, currTarget := range targetRefs {
+				if !hasSectionName(&currTarget) {
+
```
Comment (jukie) [inline internal/gatewayapi/clienttrafficpolicy.go:248]: Can you add a check here and skip processing if currTarget.Kind != passKind?

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 2 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go]: https://github.com/envoyproxy/gateway/pull/9026/changes/bb1ca6a7128f5b10783d7a910c47b7427b776b2e

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 3 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:248]: https://github.com/envoyproxy/gateway/pull/9026/changes/de7f84ae2c7517c69c057ad3e76bc0dacd04b9b9

This required re-ordering test fixture order but logic is intact, see commit message for details.

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go]: https://github.com/envoyproxy/gateway/pull/9026/changes/b628c9bb2ffd35835232f75eb5a9bc51407e2775 - applied fix and 3 new tests

**Comment by apkatsikas** [comment]
@jukie looks like the GHA for the OSV scanner is erroring on Go version

```
2026-05-22T15:28:58.9244733Z Failed to run code analysis (govulncheck) on '/github/workspace/examples/envoy-ext-auth/go.mod' because govulncheck: loading packages: err: exit status 1: stderr: go: go.mod requires go >= 1.26.3 (running go 1.26.2; GOTOOLCHAIN=local)
```

https://productionresultssa11.blob.core.windows.net/actions-results/f3abf5f9-fcf8-4c57-8235-ddd3092140cf/workflow-job-run-71a261c9-0586-5b73-b5b7-bef9d19e6dca/logs/job/job-logs.txt?rsct=text%2Fplain&se=2026-05-22T15%3A39%3A59Z&sig=%2BYM6108mvV9s0rK303laQASXeGgaXS1Qzu%2BHSi9JxD8%3D&ske=2026-05-22T18%3A56%3A30Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2026-05-22T14%3A56%3A30Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-11-05&sp=r&spr=https&sr=b&st=2026-05-22T15%3A29%3A54Z&sv=2025-11-05

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 4]
```diff
@@ -349,18 +414,71 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
+func lsPrefix(ls *gwapiv1.ListenerSet) string {
+	return fmt.Sprintf("%s/%s/", ls.Namespace, ls.Name)
+}
+
 func resolveClientTrafficPolicyTargetRef(
 	targetRef *policyTargetReferenceWithSectionName,
 	gateways map[types.NamespacedName]*policyGatewayTargetContext,
-) (*GatewayContext, *status.PolicyResolveError) {
-	// Check if the gateway exists
+	listenerSets []*gwapiv1.ListenerSet,
```
Comment (jukie) [inline internal/gatewayapi/clienttrafficpolicy.go]: Please make `listenerSets` a map and precompute similar to [how gateways are computed](https://github.com/envoyproxy/gateway/blob/9aaac883819124e12694d67ecf85fcbd2cd7925a/internal/gatewayapi/clienttrafficpolicy.go#L74-L78). Currently would cause an extra loop over all `listenerSets` per-CTP.

File: internal/gatewayapi/clienttrafficpolicy.go
Comment (jukie) [inline internal/gatewayapi/clienttrafficpolicy.go]: Is this section still accurate once LS support is added? `attachedToGateway=false` gets passed for LS-wide and then this code will treat any same-port listeners (for example an LS with two HTTP listeners on the same port) as  a failure which I don't _think_ is desired.

**Comment by jukie** [comment]
Can you also read through [GEP-1713](https://gateway-api.sigs.k8s.io/geps/gep-1713/ ) and make sure conflicts/override behavior is being handled correctly? For example we may need some additional handling in listener.go or validate.go to filter out `l.isFromListenerSet()`.

An example that may be useful to walk through is two CTPs targeting the same resulting listerner(s). CTP A targeting a ListernerSet, and CTP B targeting a Gateway. For each scenario also consider sectionName handling.

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 4 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go]: https://github.com/envoyproxy/gateway/pull/9026/changes/7ef92e91dfafacfd68a2d44ca20e0537c1825d11

File: internal/gatewayapi/clienttrafficpolicy.go
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go]: This section was not accurate once LS support was added. Addressed in https://github.com/envoyproxy/gateway/pull/9026/changes/79e9a45167adc5ac2a9f07f4ff45abb6969144f8.

**Comment by apkatsikas** [comment]
> Can you also read through [GEP-1713](https://gateway-api.sigs.k8s.io/geps/gep-1713/) and make sure conflicts/override behavior is being handled correctly? For example we may need some additional handling in listener.go or validate.go to filter out `l.isFromListenerSet()`.
> 
> An example that may be useful to walk through is two CTPs targeting the same resulting listerner(s). CTP A targeting a ListernerSet, and CTP B targeting a Gateway. For each scenario also consider sectionName handling.

I reviewed the conflict/override behavior for CTPs targeting ListenerSets vs Gateways. The existing tests already cover:
- `clienttrafficpolicy-listenerset-wide-beats-gateway-wide` and `clienttrafficpolicy-listenerset-wide-beats-gateway-wide-gw-listed-first` — LS-wide CTP beats gateway-wide regardless of input order, gateway-wide gets Overridden for all LS listeners
- `clienttrafficpolicy-gateway-listenerset-same-listener-name-sectionname` — LS sectionName CTP and gateway sectionName CTP with the same listener name do not conflict with each other

Added two missing sectionName variants in commit https://github.com/envoyproxy/gateway/pull/9026/changes/5f692239067ebb24b0bf9e183be9afe38f7d7fdc:
- `clienttrafficpolicy-listenerset-sectionname-beats-gateway-wide` — LS sectionName CTP claims one listener, gateway-wide gets Overridden for only that listener but still applies to the remaining gateway and LS listeners (verified via xdsIR keepalive config per listener)
- `clienttrafficpolicy-listenerset-wide-no-conflict-with-gateway-sectionname` — LS-wide CTP and gateway sectionName CTP target non-overlapping listeners, both accepted with neither marked Overridden

Is there a specific scenario you had in mind that isn't covered by the above?

File: internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+      name: ext-listeners
+    tcpKeepalive:
+      probes: 5
+  status:
+    ancestors:
+    - ancestorRef:
+        group: gateway.networking.k8s.io
+        kind: Gateway
+        name: composite-gateway
+        namespace: envoy-gateway
```
Comment (jukie) [inline internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml:20]: From https://gateway-api.sigs.k8s.io/geps/gep-713:
> This is an Ancestor status rather than a Parent status, as in the Route status, because, for Policy Attachment, the relevant object may or may not be the direct parent.

> For example, BackendTLSPolicy directly attaches to a Service, which may be included in multiple Routes, in multiple Gateways. However, for many implementations, the status of the BackendTLSPolicy will be different only at the Gateway level, so Gateway is the relevant Ancestor for the status. Each Gateway that has a Route that includes a backend with an attached BackendTLSPolicy SHOULD have a separate PolicyAncestorStatus section in the BackendTLSPolicy’s status.ancestors stanza, which mandates that entries must be distinct using the combination of the AncestorRef and the ControllerName fields as a key. See [GEP-1897](https://gateway-api.sigs.k8s.io/geps/gep-1897/) for the exact details.

For routes we set `ancestorRef` to the ListenerSet and in my opinion we should do the same here but I'm open to a discussion on it.

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 6]
```diff
@@ -349,18 +431,65 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
+
+		parentNamespace := NamespaceDerefOr(ls.Spec.ParentRef.Namespace, ls.Namespace)
+		gatewayKey := types.NamespacedName{
+			Namespace: parentNamespace,
+			Name:      string(ls.Spec.ParentRef.Name),
+		}
+		gateway, ok := gateways[gatewayKey]
+		if !ok {
+			return nil, nil
+		}
```
Comment (jukie) [inline internal/gatewayapi/clienttrafficpolicy.go:476]: Can we surface an error or status on the CTP when the ListenerSet wasn't accepted by its parent Gateway (e.g. excluded by `allowedListeners.namespaces.from`/`ParentNotAccepted`)?

processListenerSet emits Accepted: False on the ListenerSet in those scenarios, but this block would still produce `Accepted: True` on the CTP. I think we should reject the policy attachment instead.

CC @envoyproxy/gateway-maintainers please correct me if I'm wrong on the semantics here.

**Comment by jukie** [comment]
> Is there a specific scenario you had in mind that isn't covered by the above?

Thanks for confirming the logic, what you have above looks right to me.

I'll give this another deep look but overall looks good once the last review comments are addressed.

File: internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml (hunk 5 above)
Comment (apkatsikas) [reply internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml:20]: I agree with your opinion here. To be both consistent with the behavior of how EG handles routes, and to satisfy with how I am interpreting GEP-713 - a ListenerSet may have a different acceptance status than its parent Gateway, and the ListenerSet status is more directly relevant to the Policy in this case than the Gateway. Therefore, I would like to change this PR to derive the ancestor ref for the policy dynamically depending upon the type of the target ref (ListenerSet or Gateway), which will involve adding a new function in addition to getAncestorRefForPolicy specifically for ListenerSets, so we don't have to touch existing xPolicy callers that do not yet support LS targets (like backend traffic policy, for instance). This will also require updating the existing tests, as the golden config outputs will change for those instances where the CTP is targetting a ListenerSet, and I will also want to include a "mixed type" edge case where a single CTP targets both a Gateway and a ListenerSet — I expect this to produce two distinct ancestor entries (one Kind: Gateway, one Kind: ListenerSet), where the Gateway entry carries both Accepted and Overridden conditions (since the more-specific ListenerSet targetRef claims those listeners first), and the ListenerSet entry is simply Accepted. @jukie how does this sound?

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 6 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: I will let the maintainers weigh in as well, but this makes sense to me. After we confirm the LS is in the map, we can check if it carries a not accepted status condition, and in that case return the PolicyResolveError, and add a test case where a CTP targets a ListenerSet that has Accepted: False (e.g. due to allowedListeners namespace restriction), and the expected output shows the CTP ancestor entry with a resolve error rather than Accepted: True.

File: internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml (hunk 5 above)
Comment (jukie) [reply internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml:20]: Yeah this sounds great, thanks.

Comment (apkatsikas) [reply internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-wide-beats-gateway-wide.out.yaml:20]: https://github.com/envoyproxy/gateway/pull/9026/changes/de1f69045fbb33ec8da58d3bec1b988bd4c94d63

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 6 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: @zhaohuabing any thoughts here?

Comment (zhaohuabing) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: I think this should probably align with the existing CTP behavior for Gateway listeners. Today, if a CTP targets an existing Gateway listener that is not accepted, the policy resolution only checks that the section exists and does not reject based on the listener’s Accepted condition. The policy can still surface Accepted=True even though there may be no effective listener IR to apply to.

Given that, I think changing ListenerSet-targeted CTPs to reject when the ListenerSet has Accepted=False might be reasonable, but it is a broader semantic change from the current CTP behavior. It may be better to keep this PR consistent with the existing behavior and handle status for rejected Gateway listeners and rejected ListenerSets together in a follow-up PR.

Comment (zhaohuabing) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: Thinking about this more, I wonder if this should be modeled separately from the `Accepted ` condition. If the target ListenerSet/listener exists, the CTP may be semantically accepted as attached to that target, even if it has no data-plane effect because the target listener/ListenerSet itself is not accepted or programmed. 

This is broader than this PR, since CTP does not currently define or use a `Programmed` policy condition. I think it would be better to track this in a separate issue and discuss the best solution there.

Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: Thanks @zhaohuabing, I created https://github.com/envoyproxy/gateway/issues/9187 to track. @jukie any remaining issues to address?

Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: @jukie checking in again.

**Comment by zhaohuabing** [comment]
@codex review

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 7]
```diff
@@ -197,144 +227,206 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
-				}
+	// Policy with no section set (targeting all sections of a Gateway or ListenerSet).
+	// ListenerSet-wide policies are processed before Gateway-wide so that the more specific
+	// target always wins regardless of input order.
+	for _, passKind := range []gwapiv1.Kind{resource.KindListenerSet, resource.KindGateway} {
+		for i, currPolicy := range clientTrafficPolicies {
+			policyName := utils.NamespacedName(currPolicy)
+			targetRefs := resolvePolicyTargets(
+				currPolicy.Spec.PolicyTargetReferences,
+				gateways,
```
Comment (zhaohuabing) [inline internal/gatewayapi/clienttrafficpolicy.go:238]: Should we also pass XListenerSets into `resolvePolicyTargets`?
After fixing this, can we add a test yaml file to cover the selector case?

File: internal/gatewayapi/testdata/clienttrafficpolicy-gateway-listenerset-same-listener-name-sectionname.in.yaml [hunk 8]
```diff
@@ -0,0 +1,61 @@
+clientTrafficPolicies:
```
Comment (zhaohuabing) [inline internal/gatewayapi/testdata/clienttrafficpolicy-gateway-listenerset-same-listener-name-sectionname.in.yaml:1]: Non-blocking nit:

It would be helpful for reviewers and future readers if we could add a short comment at the top of each test file describing what it covers.

Comment (apkatsikas) [reply internal/gatewayapi/testdata/clienttrafficpolicy-gateway-listenerset-same-listener-name-sectionname.in.yaml:1]: It would be nice if it could somehow be a part of the golden files yaml structure themselves, so when the test fails it emits the scenario name or context in the output. I agree the file names are not sufficient information on the test intent.

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 7 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:238]: Yes, the selector case was not on my radar, thanks for pointing it out.

https://github.com/envoyproxy/gateway/pull/9026/changes/93a763fcc8d3a1a95675b6d443fe7a841626c369

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 9]
```diff
@@ -121,8 +145,11 @@ func (t *Translator) ProcessClientTrafficPolicies(
 
 				// Check if another policy targeting the same section exists
 				section := string(*(targetRef.SectionName))
-				s, ok := policyMap[key]
-				if ok && s.Has(section) {
+				sectionKey := section
+				if targetRef.Kind == resource.KindListenerSet {
+					sectionKey = fmt.Sprintf("%s%s", lsPrefix(ls), section)
+				}
```
[the remark this answers is not in the document]
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:151]: If I am understanding this correctly, the scenario is as follows:

1. User creates a Client Traffic Policy which targets a Gateway with section name of `ext-http-1`, but the Gateway does not have a listener called `ext-http-1`
2. A Listener Set attached to the Gateway has a listener called `ext-http-1`

At this point, I feel the user is misapplying the policy. They meant to target a Listener Set.

Continuing on:

3. Another Client Traffic Policy is being applied Listener Set wide (no section name). It re-attaches to the same Listener Set listener, so this one ends up beating out the section name and the policy is double applied.

For a user who did this, it may be surprising, but I don't consider this a real bug. It is highly theoretical and based on a faulty configuration. I don't believe the code should concern itself with this.

File: internal/gatewayapi/testdata/clienttrafficpolicy-gateway-listenerset-same-listener-name-sectionname.in.yaml (hunk 8 above)
Comment (apkatsikas) [reply internal/gatewayapi/testdata/clienttrafficpolicy-gateway-listenerset-same-listener-name-sectionname.in.yaml:1]: In https://github.com/envoyproxy/gateway/commit/93a763fcc8d3a1a95675b6d443fe7a841626c369 I add comments for those new test files.

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 10]
```diff
@@ -39,9 +39,9 @@ type ClientTrafficPolicy struct {
...
-// +kubebuilder:validation:XValidation:rule="has(self.targetRef) ? self.targetRef.kind == 'Gateway' : true", message="this policy can only have a targetRef.kind of Gateway"
+// +kubebuilder:validation:XValidation:rule="has(self.targetRef) ? self.targetRef.kind in ['Gateway', 'ListenerSet'] : true", message="this policy can only have a targetRef.kind of Gateway or ListenerSet"
 // +kubebuilder:validation:XValidation:rule="has(self.targetRefs) ? self.targetRefs.all(ref, ref.group == 'gateway.networking.k8s.io') : true", message="this policy can only have a targetRefs[*].group of gateway.networking.k8s.io"
-// +kubebuilder:validation:XValidation:rule="has(self.targetRefs) ? self.targetRefs.all(ref, ref.kind == 'Gateway') : true", message="this policy can only have a targetRefs[*].kind of Gateway"
+// +kubebuilder:validation:XValidation:rule="has(self.targetRefs) ? self.targetRefs.all(ref, ref.kind in ['Gateway', 'ListenerSet']) : true", message="this policy can only have a targetRefs[*].kind of Gateway or ListenerSet"
```
[the remark this answers is not in the document]
Comment (apkatsikas) [reply api/v1alpha1/clienttrafficpolicy_types.go:44]: https://github.com/envoyproxy/gateway/pull/9026/changes/7025aec64a341b1862930bbff8dea8eac2dc5979

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 11]
```diff
@@ -197,144 +227,206 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
 					} else {
-						if shouldDisableHTTP3ForClientValidation(policy, gwXdsIR.GetHTTPListener(irListenerName(l))) {
-							http3DisabledListeners = append(http3DisabledListeners, string(l.Name))
+						// LS-claimed sections have a scoped prefix (ns/name/listenerName);
+						// strip it so the display name is the bare listener name.
+						for _, claimedSection := range claimedSections[gatewayKey].UnsortedList() {
+							slashIndex := strings.LastIndex(claimedSection, "/")
+							if slashIndex >= 0 {
+								claimedSection = claimedSection[slashIndex+1:]
+							}
```
[the remark this answers is not in the document]
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:335]: I am intentionally using bare section names in the override message rather than scoped prefixes, as the scoped prefix is an internal implementation detail which may not be obvious to users in the current form. In the case where two ListenerSets share a listener name, both entries are preserved in the message. This is accurate since both are genuinely overriding the policy. A follow-up could provide richer disambiguation (e.g. including the ListenerSet name), but that's out of scope for this PR.

**Comment by zhaohuabing** [comment]
@codex review

**Comment by zhaohuabing** [comment]
Can we add an e2e test for CTP targeting ListenerSet?

File: internal/gatewayapi/helpers.go [hunk 12]
```diff
@@ -465,6 +465,15 @@ func extractGatewayNameFromListener(listenerName string) string {
 	return listenerName
 }
 
+func extractListenerSetPrefixFromListener(listenerName string) string {
+	parts := strings.Split(listenerName, "/")
+	if len(parts) >= 4 {
+		return fmt.Sprintf("%s/%s/%s/%s", parts[0], parts[1], parts[2], parts[3])
```
[the remark this answers is not in the document]
Comment (apkatsikas) [reply internal/gatewayapi/helpers.go]: https://github.com/envoyproxy/gateway/pull/9026/changes/a981dd4b7a8d6e426a349d208a74380feb6fe839

existing bug for CTP targetting Gateway ID-ed here https://github.com/envoyproxy/gateway/issues/9226

**Comment by apkatsikas** [comment]
> Can we add an e2e test for CTP targeting ListenerSet?

https://github.com/envoyproxy/gateway/pull/9026/changes/079cb5c4254a74b03e40d95784a946fd24d683e7

Added a couple more follow up commits for stability.

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 6 above)
Comment (apkatsikas) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: Hi @jukie - any remaining feedback or concerns on this?

**Comment by zhaohuabing** [comment]
@codex review

**Review by zhaohuabing — COMMENTED** [review]
This PR adds 30 new golden YAML files for the Gateway API tests. Are all of them necessary?

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 13]
```diff
@@ -197,144 +227,207 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
+								http3DisabledListeners = append(http3DisabledListeners, string(l.Name))
+							}
+							if err := t.translateClientTrafficPolicyForListener(policy, l, xdsIR, infraIR, resources); err != nil {
+								errs = errors.Join(errs, err)
+							}
+							if policyTargetsLS {
+								if claimedSections[gatewayKey] == nil {
+									claimedSections[gatewayKey] = sets.New[string]()
+								}
+								claimedSections[gatewayKey].Insert(fmt.Sprintf("%s%s", lsPrefix(ls), string(l.Name)))
```
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply internal/gatewayapi/clienttrafficpolicy.go:403]: The more specific policy should only override when it is accepted/effective; if it fails validation, allowing the Gateway-wide policy to apply as fallback seems reasonable.

**Review by zhaohuabing — DISMISSED** [review]
LGTM thanks!

**Comment by apkatsikas** [comment]
> This PR adds 30 new golden YAML files for the Gateway API tests. Are all of them necessary?

I believe they are all useful. The majority of them were added 1 or 2 at a time as changes were requested and PR feedback came in. Each time the code was changed to address a bug or invariant that was not being enforced, I added a test to ensure the fix worked as intended.

I think the new practice of adding the comment at the top of the test file to describe the purpose of the test, in the future, can help us decide at the end of a review if some tests can be removed at the end.

File: internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-mixed-targetrefs.out.yaml [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+        type: Accepted
+      controllerName: gateway.envoyproxy.io/gatewayclass-controller
+    - ancestorRef:
+        group: gateway.networking.k8s.io
+        kind: Gateway
+        name: composite-gateway
+        namespace: envoy-gateway
+      conditions:
+      - lastTransitionTime: null
+        message: There are existing ClientTrafficPolicies that are overriding these
```
Comment (kkk777-7) [inline internal/gatewayapi/testdata/clienttrafficpolicy-listenerset-mixed-targetrefs.out.yaml:38]: nits:
Would the following message provide a better user experience?
```
message: There are existing ClientTrafficPolicies that are overriding these sections [envoy-gateway/ext-listeners/ext-http-1 envoy-gateway/ext-listeners/ext-http-2]
```

**Review by kkk777-7 — DISMISSED** [review]

**Comment by kkk777-7** [comment]
LGTM, thanks!

I left one nits comment(not-blocking). we could do follow-up PR.
It would be great to update docs in follow-up PR.
https://gateway.envoyproxy.io/docs/concepts/gateway_api_extensions/client-traffic-policy/#clienttrafficpolicy-in-envoy-gateway
e.g. Policies attached to a Gateway affect ListenerSets.

**Review by jukie — APPROVED** [review]

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 15]
```diff
@@ -197,144 +227,207 @@ func (t *Translator) ProcessClientTrafficPolicies(
...
+			targetRefs := resolvePolicyTargetsForGatewayAndListenerSet(
+				currPolicy.Spec.PolicyTargetReferences,
+				gateways,
+				resources.ListenerSets,
+				resources.ReferenceGrants,
+				egv1a1.GroupName,
+				egv1a1.KindClientTrafficPolicy,
+				currPolicy.Namespace,
+				t.GetNamespace,
+			)
```
Comment (jukie) [inline internal/gatewayapi/clienttrafficpolicy.go:245]: Non-blocking: one last perf observation is that we we're processing every CTP twice, once for ListenerSets and a second for Gateways. I think we can split the `clientTrafficPolicies` loop and resolve policy targets before the `passKind` loop.

Something like this:
```
policyTargets := make([][]policyTargetReferenceWithSectionName, len(clientTrafficPolicies))
for i, p := range clientTrafficPolicies {
    policyTargets[i] = resolvePolicyTargetsForGatewayAndListenerSet(
        p.Spec.PolicyTargetReferences, gateways, resources.ListenerSets,
        resources.ReferenceGrants, egv1a1.GroupName, egv1a1.KindClientTrafficPolicy,
        p.Namespace, t.GetNamespace,
    )
}

for _, passKind := range []gwapiv1.Kind{resource.KindListenerSet, resource.KindGateway} {
    for i := range clientTrafficPolicies {
        for _, currTarget := range policyTargets[i] {
            ...
        }
    }
}
```

**Review by kkk777-7 — APPROVED** [review]

File: internal/gatewayapi/clienttrafficpolicy.go (hunk 6 above)
Comment (jukie) [reply internal/gatewayapi/clienttrafficpolicy.go:476]: No concerns. Sounds good and can be addressed with #9187

**Review by zirain — APPROVED** [review]
